graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/AbstractGraphqlService.java

```diff
@@ -90,6 +97,42 @@ protected HttpResponse doPost(ServiceRequestContext ctx, HttpRequest request) th
             return unsupportedMediaType();
         }
 
+        // See https://github.com/jaydenseric/graphql-multipart-request-spec/blob/master/readme.md
+        if (contentType.is(MediaType.MULTIPART_FORM_DATA)) {
+            return HttpResponse.from(FileAggregatedMultipart.aggregateMultipart(ctx, request)
+                                                            .thenApply(multipart -> {
+                try {
+                    final ListMultimap<String, String> multipartParams = multipart.params();
+                    final String operationsParam = getValueFromMultipartParam("operations", multipartParams);
+                    final String mapParam = getValueFromMultipartParam("map", multipartParams);
+
+                    final Map<String, Object> operations = parseJsonString(operationsParam, JSON_MAP);
+                    final Map<String, List<String>> map = parseJsonString(mapParam, MAP_PARAM);
+                    final String query = toStringFromJson("query", operations.get("query"));
```

Comment:
Now, the parsing logic is scattered in 3 places. Is it time to refactor that to make it more maintainable?
We can add some factory methods to `GraphqlRequest`. 
```java
interface GraphqlRequest {
    static GraphqlRequest parse(String json) {  ... }
    // It seeems useful to add `QueryParams.toMap()` and `QueryParams.toMultiMap()`.
    static GraphqlRequest from(QueryParams queryParams) { ... }
    static GraphqlRequest from(Map<String, Object> params) {  ... }
}
```

Multipart-data will focus on `bindMultipartVariable` to fill variables.
```java
final ListMultimap<String, String> multipartParams = multipart.params();
final String operationsParam = getValueFromMultipartParam("operations", multipartParams);
final String mapParam = getValueFromMultipartParam("map", multipartParams);

final Map<String, Object> operations = parseJsonString(operationsParam, JSON_MAP);
...
bindMultipartVariable(copiedVariables, entry.getValue(), multipartFiles);

// Validations will be performed and query and extension are extracted in the `from()`.
GraphqlRequest.from(operations);
```

But I don't see we need to handle it in this PR. @heowc How about doing this work in the following PR?

Reply:
I also agree with this. 👍 
Not only that, but `AbstractGraphqlService` itself is subject to refactoring, as it does quite a bit of work.